src/System.IO.Pipelines/SegmentEnumerator.cs

```diff
@@ -33,50 +35,55 @@ public SegmentEnumerator(ReadCursor start, ReadCursor end)
         /// Moves to the next <see cref="Memory{Byte}"/> in the <see cref="ReadableBuffer"/>
         /// </summary>
         /// <returns></returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public bool MoveNext()
         {
-            if (_segment == null)
+            var segment = _segment;
+
+            if (segment == null)
             {
                 return false;
             }
 
-            int start = _segment.Start;
-            int end = _segment.End;
+            var start = _startIndex;
+            var end = segment.End;
 
-            if (_startIndex != 0)
-            {
-                start = _startIndex;
-                _startIndex = 0;
-            }
-
-            if (_segment == _endSegment)
+            if (segment == _endSegment)
             {
                 end = _endIndex;
-            }
-
-            _current = new SegmentPart()
-            {
-                Segment = _segment,
-                Start = start,
-                End = end,
-            };
-
-            if (_segment == _endSegment)
-            {
                 _segment = null;
             }
             else
             {
-                _segment = _segment.Next;
+                _segment = segment.Next;
                 if (_segment == null)
                 {
-                    throw new InvalidOperationException("Segments ended by end was never seen");
+                    if (_endSegment != null)
+                    {
+                        ThrowEndNotSeen();
+                    }
+                }
+                else
+                {
+                    _startIndex = _segment.Start;
                 }
             }
 
+            _current = new SegmentPart()
+            {
+                Segment = segment,
+                Start = start,
+                End = end,
+            };
```

Comment:
just update `_current` rather than calling `.ctor` + updates `{ ... }`?
```csharp
_current.Segment = segment;
_current.Start = start;
_current.End = end;
```
Might be good, might not...